mesonbuild/dependencies/python.py

```diff
@@ -170,11 +170,24 @@ def __init__(self, name: str, environment: 'Environment',
         SystemDependency.__init__(self, name, environment, kwargs)
         _PythonDependencyBase.__init__(self, installation, kwargs.get('embed', False))
 
+        # link args
         if mesonlib.is_windows():
             self.find_libpy_windows(environment)
         else:
             self.find_libpy(environment)
 
+        # compile args
+        inc_paths = mesonlib.OrderedSet([
+            self.variables.get('INCLUDEPY'),
+            self.paths.get('include'),
+            self.paths.get('platinclude')])
+
+        self.compile_args += ['-I' + path for path in inc_paths if path]
+
+        # https://sourceforge.net/p/mingw-w64/mailman/message/30504611/
+        if mesonlib.is_windows() and self.get_windows_python_arch() == '64' and self.major_version == 2:
+            self.compile_args += ['-DMS_WIN64']
+
```

Comment:
Again, not related to the refactoring, but compiling Cython extension modules with GCC on Windows requires to add this to `meson.build` also for Python 3:
```meson
if host_machine.system() == 'windows' and meson.get_compiler('c').get_id() == 'gcc'
  add_project_arguments('-DMS_WIN64', language: ['c', 'cpp'])
endif
```
therefore I think the `and self.major_version == 2` guard should be dropped.

Reply:
I really have no clue how that works, but I guess that can be done... in a separate commit. ;)

@MathieuDuponchelle, do you happen to remember why you originally guarded it by version 2.x, way back in commit e1b138a21b45fe9b55f3bb7d9b1e0bbbd45a7609?

Reply:
Hey, I should have documented that part specifically but only added the thread that justified adding `-DMS_WIN64`, I have absolutely no clue at this point why I added the python 2 restriction, sorry about that.

Reply:
No big deal, looks like it may have just been due to unclear wording in the linked message...

I bet @rgommers knows a lot about the topic anyway and can confirm the correctness of such a change.

Reply:
The SciPy code comment points at https://github.com/python/cpython/issues/72454. There's a bunch of MinGW patches that are needed but were never accepted by CPython (they basically never cared much about anything but MSVC). It looks like it was very recently resubmitted and merged (https://github.com/python/cpython/pull/100137), so it should be fixed for Python `3.12.0`. So the define is needed for Python 3 up to today, and hence dropping the `== 2` sounds good to me.

And, just to be sure: if `-DMS_WIN64` is added twice (once by Meson, once by e.g. SciPy), that won't hurt will it?

Reply:
Thanks for the context -- that's wonderfully actionable and gives me a lot of confidence about what to do here.

And yes, double-defining it should be fine, the definitions are compatible so it's a no-op.